File: python/ctci/chp02_linked_list/linkedlist.py
from os import link
from re import L
from tkinter import E

from pip import main


class Node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next

    def __str__(self):
        return f"value: {str(self.value)}"


class LinkedList:
    def __init__(self):
        self.head = None

    def __str__(self):

        if self.head is None:
            print("linked list is empty")
            return

        current = self.head
        message = ""
        while current:
            message += f"{current.value}"
            if current.next is not None:
                message += f" -> "
            current = current.next
        return message

    def from_list(self, l):
        for item in l:
            self.append(item)

    def append(self, value):
        if self.head is None:
            self.head = Node(value)
            return

        current = self.head
        while current.next:
            current = current.next
        current.next = Node(value)

    def prepend(self, value):
        if self.head is None:
            self.head = Node(value)
        else:
            new_node = Node(value)
            new_node.next = self.head
            self.head = new_node

    def size(self):
        if self.head is None:
            return 0
        else:
            size = 0
            current = self.head
            while current:
                size += 1
                current = current.next
            return size

    def remove_by_index(self, remove_index):
        # check if remove_index out of range
        if remove_index > self.size() - 1:
            print("remove index out of range")
            return

        # check if linked list is empty
        if self.head is None:
            return None

        current_index = 0

        # check we should remove head
        if current_index == remove_index:
            temp = self.head
            self.head = self.head.next
            del temp
            return

        # iterate through list to index and remove Node
        current = self.head
        while current:
            if current_index + 1 == remove_index:
                temp = current.next
                current.next = current.next.next

            current = current.next
            current_index += 1

File: c#/practice/data-structures/LinkedList.cs
using System;
using System.Collections;

public class Node
{
  public int val;
  public Node next;
  public Node(int val = 0, Node next = null)
  {
    this.val = val;
    this.next = null;
  }
}

public class LinkedList
{
  private Node head;
  private int size;
  public LinkedList()
  {
    this.head = null;
  }

  public void append(int val)
  {
    if (this.head == null)
    {
      this.head = new Node(val);
      this.size++;
      return;
    }
    else
    {
      Node current = this.head;
      while (current != null && current.next != null)
      {
        current = current.next;
      }

      current.next = new Node(val);
      this.size++;
    }
  }

  public void prepend(int val)
  {
    if (this.head == null)
    {
      this.head = new Node(val);
      this.size++;
      return;
    }
    else
    {
      Node newNode = new Node(val);
      newNode.next = this.head;
      this.head = newNode;
      this.size++;
      return;
    }
  }

  public void remove(int index)
  {
    if (index < 0 || index > this.size)
    {
      return;
    }

    int n = 0;
    Node current = this.head;
    Node prev;
    prev = null;

    if (index == 0)
    {
      this.head = current.next;
      this.size--;
      return;
    }

    while (current != null)
    {
      if (n == index)
      {
        prev.next = current.next;
        this.size--;
        return;
      }
      n++;
      prev = current;
      current = current.next;
    }

  }

  public void printList()
  {
    if (this.head == null)
    {
      return;
    }

    Node current = this.head;
    Console.Write(current.val);
    current = current.next;
    while (current != null)
    {
      Console.Write("-->" + current.val);
      current = current.next;
    }

    Console.WriteLine();
  }
}

public class Program
{
  public static void Main()
  {
    LinkedList ll = new LinkedList();
    ll.append(5);
    ll.append(6);
    ll.prepend(4);
    ll.remove(1);
    ll.printList();
  }
}